design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -399,18 +412,109 @@ are the basic configurations that the tool should support.
 Example of `ksql-migrations.properties`
 ```
 # Server URL and authentication
-ksql.server.url='http://localhost:8080'
-ksql.username='user1'
-ksql.password='pass1'
+ksql.server.url=http://localhost:8080
+ksql.username=user1
+ksql.password=pass1
+
+# TLS configs, including for mTLS
+ssl.truststore.location=/path/to/truststore
+ssl.truststore.password=pass
+ssl.keystore.location=/path/to/keystore
+ssl.keystore.password=pass
+ssl.key.password=pass
 
 # Migrations details
-ksql.migrations.streamName='migration_events'
-ksql.migrations.tableName='schema_version'
-ksql.migrations.createSchemas='true'
+ksql.migrations.stream.name=migration_events
+ksql.migrations.table.name=schema_version
+ksql.migrations.stream.topic.name=default_ksql_migration_events
+ksql.migrations.table.topic.name=default_ksql_schema_version
+ksql.migrations.topic.replicas=1
 ```
 
+The default stream and table metadata topic names will be `{clusterID}ksql_{StreamOrTableName}`, where the migrations
+tool with determine the ksqlDB cluster ID by accessing the `/info` endpoint of the ksqlDB server. If desired,
+these topic names may be overridden by the user.
+
 Note: The command line options and other configurations will also be defined during the implementation.
 
+### Migration state transitions and potential race conditions
+
+Below we have two potential options for how to manage migration state. We've decided to go with the latter
+("Do not implement a transaction mechanism and accept the possibility of race conditions") for the first
+version of migrations tool as it should be uncommon (and discouraged) to have multiple users/automated jobs
+running migrations simultaneously.
+
+#### Use Kafka transactions to avoid race conditions
+
+To avoid the possibility of multiple migrations being performed at the same time, we can emulate how the
+ksqlDB command topic uses Kafka transactions to ensure that only one migration takes place at a time. 
+We'd set the producer's transactional ID based on the ksqlDB cluster's service ID (which the migrations
+tool can obtain via the `/info` endpoint of the ksqlDB server), and the flow for applying a migration would be:
+1. Begin transaction.
+1. Check `CURRENT` state is either `Migrated` or `Error` (i.e., not `Running`), by creating a Kafka consumer. Return error and abort if `Running`.
```

Comment:
> nit: all steps are numbered 1.?

That's markdown syntax for creating a numbered list. You'll see it looks fine when rendered :) https://github.com/confluentinc/ksql/blob/master/design-proposals/klip-42-schema-migrations-tool.md#use-kafka-transactions-to-avoid-race-conditions

> Also I'm wondering what's the default "transaction.timeout" config in ksqlDB? Reading data within a producer txn could take some time and may cause the txn to expire; how is that resolved e.g. in command topic today?

I don't see anywhere ksqlDB sets the config so I assume that means we're using the default of 1 minute. If we hit the timeout when submitting commands to the command topic, the request is aborted and it's up to the user to try again. This might happen if the server that received the request is not caught up on processing the command topic: https://github.com/confluentinc/ksql/blob/e2b26c59667f605f9742b0dbf9780dee17c1f58e/ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/computation/DistributingExecutor.java#L141-L142